### ISSUE TITLE
haxe server seems to fail to cache class (and its build macro) as soon as the class's file is queried once as the current `display` file 

### ISSUE THREAD

**Issue by jeremyfa** [HaxeFoundation/haxe#6795]
I am getting slower and slower code completion as my haxe project gets bigger and relies more and more on build macros. Thus I tried to understand what happens with ``--times -D macro-times -D eval-Times`` and noticed that my build macros are not cached. The project being quite complex it was very difficult to know what prevents the macros from being cached, thus I decided to create a "blank" and much simpler project with a similar build macro, to see if in that case the macro gets cached properly. I added a ``Sys.sleep()`` call in the build macro to notice very quickly if the macro is cached or not.

My results are: sometimes it does, sometimes it doesn't and I managed to reproduce the steps to make it fail.

The example project is here: https://github.com/jeremyfa/haxe-macro-cache-test and I can reproduce the issue consistently with latest vscode/vshaxe and haxe 3.4.4.

1. When I open the project folder on vscode, the haxe server is started and everything seems to be cached properly. I can type code inside the ``Main`` (``src/Main.hx``) class and code completion is instant, the build macro doesn't need to be run again, it was run once when the haxe server was started by vscode, _everything is fine_.

2. I open one of the files that depend on the build macro: ``src/proj/SomeClass.hx``. This makes vscode run diagnostics on it, and query haxe server with this same file **as the display file**, so it is parsed again, macro executed again. That makes sense as it is the file currently being viewed/edited. I leave this file untouched.

3. I get back to ``src/Main.hx`` and type code again in it: this time, code completion gets very slow because ``SomeClass``'s build macro is executed instead of using a cached version. I try again and again and everytime the build macro is executed and code completion is slow. It's like it never gets cached again, doomed to be re-evaluated each time.

4. I need to reload vscode window (restard haxe server) to get back a cached version of ``SomeClass``.

In practice, that means every time I open a file in vscode that depends on the build macro, the haxe server gets slower as the opened file is never cached anymore and server executes one more build macro at each query.

I can't fix any caching issue on the bigger project as long as this _simpler_ one remain, so I hope the reproducible example can help finding the problem 🙏

**Comment by ncannasse** [comment]
Please run your compilation server with  -v (before --wait), this should tell you the reason the file is reloaded. Also check that closing the file in VSCode bring back the speed to normal, in which case it means that everytime you trigger compilation vshaxe will save each opened file, changing their timestamps and requiring the compiler to compile them again, this would then be a vshaxe issue not a haxe one.

**Comment by jeremyfa** [comment]
[haxe-server-logs.txt](https://github.com/HaxeFoundation/haxe/files/1584044/haxe-server-logs.txt)

_Sorry if the message is long, I am trying to be very precise and detailed as it's pretty easy to lose track of what happens_

Above is the full dump of haxe server logs minus the <il> contents to make it more readable and added comments (between `==` like ``== OPEN src/Main.hx ==``) to understand what is done in the editor in parallel.

When the build macro is run, there is also a log that appears in the server ``-- EXECUTE EVENTS MACRO [ targetClass ] --``. We can see that it is executed twice when the server is generating initial cache, one for class ``proj.Entity`` and one for ``proj.SomeClass``. This is expected.

So, first I open ``src/Main.hx``, type code, it uses the cache properly, no macro is run, diagnostics are run again and use cache, that seem perfect (see below).

```
== TYPE ONE CHARACTER IN src/Main.hx ==

Processing Arguments [--cwd,/Users/jeremyfa/Developer/cachetest,build.hxml,-D,display-details,--no-output,--display,/Users/jeremyfa/Developer/cachetest/src/Main.hx@0@diagnostics]
Defines cpp=1,dce=std,display-details=1,display=diagnostics,display_details=1,haxe3=1,haxe_ver=3.404,hxcpp_api_level=331,no_copt=1,source-header=Generated by Haxe 3.4.4,sys=1,true=1
Using signature c99c535697499ca89aab4a3f5da5ec1c
Display position: /Users/jeremyfa/Developer/cachetest/src/Main.hx: 0-0
Found context 0:
{
	version = 3404;
	args = [--display, /Users/jeremyfa/Developer/cachetest/src/Main.hx@0@diagnostics, -main, Main, -cp, src, -cpp, bin, -cmd, bin/Main, -D, display-details, --no-output];
	debug = false;
	platform = cpp;
	std_path = [/usr/lib/haxe/std/, /usr/share/haxe/std/, /usr/local/lib/haxe/std/];
	class_path = [src/, /usr/lib/haxe/extraLibs/, /usr/local/lib/haxe/extraLibs/, , /usr/lib/haxe/std/cpp/_std/, /usr/share/haxe/std/cpp/_std/, /usr/local/lib/haxe/std/cpp/_std/, /usr/lib/haxe/std/, /usr/share/haxe/std/, /usr/local/lib/haxe/std/];
	defines = {true = 1, sys = 1, source-header = Generated by Haxe 3.4.4, no_copt = 1, hxcpp_api_level = 331, haxe_ver = 3.404, haxe3 = 1, display_details = 1, display-details = 1, display = diagnostics, dce = std, cpp = 1};
	defines_signature = c99c535697499ca89aab4a3f5da5ec1c;
}
 0,cpp: skipping Main
 0,cpp: parsed /Users/jeremyfa/Developer/cachetest/src/Main.hx (not cached, is display file)
Completion Response =
[]
Found context 1:
{
	version = 3404;
	args = [--display, /Users/jeremyfa/Developer/cachetest/src/Main.hx@0@diagnostics, -main, Main, -cp, src, -cpp, bin, -cmd, bin/Main, -D, display-details, --no-output];
	debug = false;
	platform = neko;
	std_path = [/usr/lib/haxe/std/, /usr/share/haxe/std/, /usr/local/lib/haxe/std/];
	class_path = [/usr/lib/haxe/std/neko/_std/, /usr/share/haxe/std/neko/_std/, /usr/local/lib/haxe/std/neko/_std/, src/, /usr/lib/haxe/extraLibs/, /usr/local/lib/haxe/extraLibs/, , /usr/lib/haxe/std/, /usr/share/haxe/std/, /usr/local/lib/haxe/std/];
	defines = {true = 1, sys = 1, source-header = Generated by Haxe 3.4.4, no_copt = 1, neko = 1, macro = 1, hxcpp_api_level = 331, haxe_ver = 3.404, haxe3 = 1, display_details = 1, display-details = 1, display = diagnostics, dce = std};
	defines_signature = eef8ab5b12c22bf6d68bc9a1cff9cf59;
}
 1,  n: found 40 directories
 0,cpp: found 45 directories
Stats = 1 files, 1 classes, 1 methods, 0 macros
Time spent : 0.000s
Processing Arguments [--cwd,/Users/jeremyfa/Developer/cachetest,build.hxml,-D,display-details,--no-output,--display,/Users/jeremyfa/Developer/cachetest/src/Main.hx@542@toplevel,-D,display-stdin]
Defines cpp=1,dce=std,display-details=1,display-stdin=1,display=toplevel,display_details=1,display_stdin=1,haxe3=1,haxe_ver=3.404,hxcpp_api_level=331,no_copt=1,source-header=Generated by Haxe 3.4.4,sys=1,true=1
Using signature c99c535697499ca89aab4a3f5da5ec1c
Display position: /Users/jeremyfa/Developer/cachetest/src/Main.hx: 542-542
 0,cpp: skipping Main
Completion Response =
<il>...</il>
Stats = 9 files, 51 classes, 268 methods, 0 macros
Time spent : 0.030s
```

Once I open ``src/proj/SomeClass.hx``, the server parses the file as it is the _current_ display file, as we can see in the logs:

```
Display position: /Users/jeremyfa/Developer/cachetest/src/proj/SomeClass.hx: 0-0
 0,cpp: skipping Main
 0,cpp: skipping proj.SomeClass
 0,cpp: parsed /Users/jeremyfa/Developer/cachetest/src/proj/SomeClass.hx (not cached, is display file)
Completion Response =
[]
Stats = 2 files, 2 classes, 7 methods, 1 macros
Time spent : 0.010s
Haxe print:
-- EXECUTE EVENTS MACRO [ proj.SomeClass ] --
```

The macro is executed again as the file is not cached. That's ok as well, it's the current file.

But then, I get back to ``src/Main.hx`` type code again, and the macro is still executed for ``proj.SomeClass``. The haxe server doesn't seem to say anything about it being reparsed or not though. But the macro is still re-run (see below the ``-- EXECUTE EVENTS MACRO [ proj.SomeClass ] --``)

```
== GET BACK TO src/Main.hx ==

Processing Arguments [--cwd,/Users/jeremyfa/Developer/cachetest,build.hxml,-D,display-details,--no-output,--display,/Users/jeremyfa/Developer/cachetest/src/Main.hx@0@diagnostics]
Defines cpp=1,dce=std,display-details=1,display=diagnostics,display_details=1,haxe3=1,haxe_ver=3.404,hxcpp_api_level=331,no_copt=1,source-header=Generated by Haxe 3.4.4,sys=1,true=1
Using signature c99c535697499ca89aab4a3f5da5ec1c
Display position: /Users/jeremyfa/Developer/cachetest/src/Main.hx: 0-0
 0,cpp: skipping Main
 0,cpp: parsed /Users/jeremyfa/Developer/cachetest/src/Main.hx (not cached, is display file)
 0,cpp: skipping proj.SomeClass
Completion Response =
[]
Stats = 2 files, 2 classes, 7 methods, 1 macros
Time spent : 0.010s
Haxe print:
-- EXECUTE EVENTS MACRO [ proj.SomeClass ] --
Haxe print:


Processing Arguments [--cwd,/Users/jeremyfa/Developer/cachetest,build.hxml,-D,display-details,--no-output,--display,/Users/jeremyfa/Developer/cachetest/src/Main.hx@0@diagnostics]
Defines cpp=1,dce=std,display-details=1,display=diagnostics,display_details=1,haxe3=1,haxe_ver=3.404,hxcpp_api_level=331,no_copt=1,source-header=Generated by Haxe 3.4.4,sys=1,true=1
Using signature c99c535697499ca89aab4a3f5da5ec1c
Display position: /Users/jeremyfa/Developer/cachetest/src/Main.hx: 0-0
 0,cpp: skipping Main
 0,cpp: parsed /Users/jeremyfa/Developer/cachetest/src/Main.hx (not cached, is display file)
 0,cpp: skipping proj.SomeClass
Completion Response =
[]
Stats = 1 files, 2 classes, 7 methods, 1 macros
Time spent : 0.010s
Haxe print:
-- EXECUTE EVENTS MACRO [ proj.SomeClass ] --
Haxe print:

== TYPE A CHARACTER IN src/Main.hx ==

Processing Arguments [--cwd,/Users/jeremyfa/Developer/cachetest,build.hxml,-D,display-details,--no-output,--display,/Users/jeremyfa/Developer/cachetest/src/Main.hx@542@toplevel,-D,display-stdin]
Defines cpp=1,dce=std,display-details=1,display-stdin=1,display=toplevel,display_details=1,display_stdin=1,haxe3=1,haxe_ver=3.404,hxcpp_api_level=331,no_copt=1,source-header=Generated by Haxe 3.4.4,sys=1,true=1
Using signature c99c535697499ca89aab4a3f5da5ec1c
Display position: /Users/jeremyfa/Developer/cachetest/src/Main.hx: 542-542
 0,cpp: skipping Main
 0,cpp: skipping proj.SomeClass
Completion Response = <il>...</il>
Stats = 1 files, 52 classes, 274 methods, 1 macros
Time spent : 0.020s
Haxe print:
-- EXECUTE EVENTS MACRO [ proj.SomeClass ] --
Haxe print:
```

Closing ``src/proj/SomeClass.hx`` doesn't seem to change anything either. Macro is run everytime, even though haxe server doesn't say it parsed SomeClass.hx again.

It doesn't look like vshaxe is saving anything. I know it would save files when I run _build_, but it is not the case here as I am not building the project explicitly, just opening files, typing code to get code completion and and diagnostics, and letting vshaxe query the server for that.

To me it looks more like an issue in the haxe server than in vshaxe, I guess?

**Comment by nadako** [comment]
> everytime you trigger compilation vshaxe will save each opened file

vshaxe doesn't do that, it uses --display stdio (only for the current file) instead

**Comment by ncannasse** [comment]
There is something that you should keep in mind when dealing with compilation server, is that the server cache is only kept in memory when a full build has succeeded.

So when you modify your macro file, it will not be back into the cache until another full compilation has terminated with 0 errors.

This is a known limitation, mainly because it is hard with type inference to pinpoint exactly which modules should be validated or rejected in case of an error, and also because of the lazy typing aspect of the compiler which might cache some modules which are not fully typed yet.

Can you try again by doing a full compilation after having edit the macro file and see if that solves the problem?

**Comment by jeremyfa** [comment]
That makes sense, but here I don't even need to edit the macro file (or class that depends on the macro) to make the cache break. Just opening the file (and having vshaxe run a ``--display`` on this opened file) is enough to make the server never use the cached version again. The file is not modified, not saved, just provided once to the server with --display stdio. I would expect that after that, the server would still reuse the cached version as long as it is not the "current display" file and it was not modified?

I'm not sure how I can do a full compilation using the same haxe server as vshaxe's (@nadako does the default haxe build command will use the haxe server vshaxe created for completion and diagnostics? And if not, is there a way to do so?).

All I can say is that if I restart the haxe server, cache on the macro works again IF (and only IF) the macro-dependent file is not open in the editor.

**Comment by Simn** [comment]
And yet another issue of yours which I _think_ is resolved.
